### ISSUE TITLE
Please modify description of a dashboard and create a new dashboard for unique CVEs

### ISSUE THREAD

**Issue by afeddersen** [chainguard-dev/rumble#74]
As we discussed in Slack...

- The total Grype and Trivy vuln findings count for Chainguard Images is accurate but because a package might be included in multiple images we’re likely counting the occurrence multiple times.  Would you mind adding a note to this chart to make sure everyone is aware of this detail?  
- Can you create a new timeseries chart that shows the unique count of CVEs over time for Grype and Trivy?

Thank you!

**Comment by jspeed-meyers** [comment]
I'll work on this on Friday, June 9 and report back. This should be just some button mashing :)

**Comment by jspeed-meyers** [comment]
@afeddersen, good news and bad news.

**Good news**: Modifying the description of the existing [dashboard](https://lookerstudio.google.com/s/oNOJeJjjRZU) is easy. See screenshot below.

![Screenshot 2023-06-09 at 10 44 29 AM](https://github.com/chainguard-dev/rumble/assets/54914994/06c61807-5794-4488-ab91-06f5a733a29d)

Does that do the trick there?

**Bad news**: The current data collection system does NOT collect data at the CVE level, at least not in a format that is readily analyzed via Google data studio. Translation: Creating a unique CVE count over time dashboard is not currently possible.

*Explanation*: `rumble` currently collects aggregate CVE data for each image from both `grype` and from `trivy`. So the database includes an entry each day for each image for each scanner with total number of vulnerabilities, total number of critical vulnerabilities, etc. To count unique CVEs though, the aggregate data is inadequate though.

*The semi-good news*: Adapting `rumble` to also collect detailed CVE data should be possible. I could imagine creating a new Google BigQuery database and adapting the rumble code to ship a new record for every CVE associated with an image (from each scanner). A dashboard count then show the number of unique CVEs for a given day. (There is a dragon or two here, but this would still be a substantial improvement over the current graph.)

Of course, this would involve some engineering work. @jdolitsky, not to volunteer him but to recognize his expertise and talent, is the engineer that, as far as I am concerned, created the current `rumble` architecture, wrote the code, has nursed it with me (I'm mainly a consumer, more like a bad PM with some bad data skills), and could provide an estimate of technical options and potential timelines. In short, I defer to you and the Images team about whether pursuing this feature is worth the engineering expense. I personally would find that data valuable for future DevRel for Images activities (blog posts and presentations and such).

What do you think? I apologize for mistakenly thinking this was simply a button-mashing exercise.

**Comment by jdolitsky** [comment]
willing!

**Comment by jspeed-meyers** [comment]
@jdolitsky, I'll put a time on your calendar for hacking on this! Thank you.

**Comment by luhring** [comment]
> And there might be a gotcha or two on filtering on that regex [...]

If we're trying to get at "unique vulnerabilities matched to packages among all our images", then yes, I would not restrict vulnerability matches to those whose vulnerability ID matches that "CVE" regex.

Extra context, **ignore if not useful**: vulnerabilities might not be unique _even within a single image_ for the following reasons:
- a given vulnerability can be matched to what appears to be the same package multiple times — e.g. the same Go module but in two different Go binaries at two different filepaths — where remediation must be done in all locations, not just once per vulnerability ID
- a given vulnerability can be matched to multiple package types, such as when there's an APK package and a Python package with the same name, both of which describe the same underlying code

Happy to join a call if I can help with knocking this out in any way!

**Comment by jspeed-meyers** [comment]
@luhring, gosh, thank you for this context. I think that Josh and I will first hack a little, and then we'll come back to this thread with some proposals for you to review. Thank you again!

**Comment by jspeed-meyers** [comment]
Josh and I had to postpone this meeting until next week give the images summit. Will report back then.

**Comment by luhring** [comment]
After spending time at the images summit on data queries and vuln remediation from that data, I think what we want isn't unqiue CVEs, it's **_unique package-to-vulnerability matches_**.

Reasons:

1. First, as noted above, we don't just want CVEs, we want all kinds of vulnerabilities that are showing up in customers' scanning.
2. Second, if `1` vulnerability is matched to `5` Wolfi packages, we don't have `1` remediation action item, we have `5`.

And, to arrive at data in this format, I think it'd be helpful to be able to filter package-vulnerability-matches based on the image in which the match occurred — and to filter on images based on which customer(s) we're prioritizing (and thus prioritizing their images).

Curious for your thoughts @afeddersen and @jspeed-meyers!

**Comment by jspeed-meyers** [comment]
> I think what we want isn't unqiue [sic] CVEs, it's unique package-to-vulnerability matches.

Agreed.

> Second, if 1 vulnerability is matched to 5 Wolfi packages, we don't have 1 remediation action item, we have 5.

Yeah, this perspective makes sense to me. I wonder if @afeddersen had another perspective in mind.

> and to filter on images based on which customer(s) we're prioritizing (and thus prioritizing their images).

The `rumble` data doesn't understand the customer to image mapping (at least yet), so there would have to be some SQL-fu, assuming that mapping exists somewhere else.

Thank for you the clear thinking!

**Comment by jspeed-meyers** [comment]
@jdolitsky has started exploring collecting data CVE by CVE with this WIP PR: https://github.com/chainguard-dev/rumble/pull/79

Me knowing Josh, that should be all doable. Once that PR lands and there is CVE by CVE data, it's off to the races with Looker Studio.

Big thanks to @luhring for helping Josh and me think this through.

**Comment by jspeed-meyers** [comment]
https://github.com/chainguard-dev/rumble/pull/79 is no longer WIP. @luhring and I are providing a review.

**Comment by jdolitsky** [comment]
That change is ready, going to sit down with @jspeed-meyers to land it on Monday July 26

Here is a preview of what the new table looks like:
![image](https://github.com/chainguard-dev/rumble/assets/393494/edb5aef3-8d08-46a8-8ff9-c6d520fc207b)

**Comment by jdolitsky** [comment]
This data should be available after tonights/tomorrows scan

**Comment by jspeed-meyers** [comment]
@jdolitsky, do you think it's time for me to add a graph to looker studio that shows the unique CVE count over time? I think it is, but I wanted to make sure I'm not missing something. I could wait until Monday so we have four days of data.

**Comment by jspeed-meyers** [comment]
@jdolitsky, friendly ping :)

@afeddersen, do you still want that graph of unique CVEs over time? I'm assuming so.

**Comment by luhring** [comment]
Unless we're sure we need _"unique CVEs"_, I'd like to recommend the graphs we develop proceed with just the raw image vulnerability matches.

My concern is that "unique CVEs" might not gain us clarity, and we'd lose some important information as we're working on observing the fire and putting out the fire. I think the "unique CVEs" metric would correlate with our progress, but only at a macro level — and on a week to week basis, it wouldn't necessarily reveal the progress we've made or the customer-visible improvements to the situation.

**Comment by afeddersen** [comment]
I think there needs to be 1-2 metrics that help us communicate our status and the progress that we're making in this area.

As a customer, I'd want to know the following in advance:

1. How many CVEs are in the images I just bought from Chainguard?  (i.e., as measured by Grype but at some point it would be great to know the count from _all_ of the scanners.
2. How many of these CVEs are not being tracked by Chainguard?

I don't know if this is realistic, but it would be great if the total count across the entire catalog was consistently < 10.  And it would be ideal if the total count always matched the number being tracked by Chainguard in the advisory dataset.

*Important:*  I'm describing the ideal metric(s) but I'm not sure if Rumble needs to be the system that produces them.  I really appreciate all of the work that you've done on this issue but we should take into account the remaining effort/time before continuing to do more.  I think we have at least 2 systems with some amount of overlap so we should consider our options.  :)

**Comment by jspeed-meyers** [comment]
> I think we have at least 2 systems with some amount of overlap so we should consider our options. :)

I am truly indifferent to what system should capture this data. I love, love, love @rawlingsj's insights works. If this metric should come from insights, great! If `rumble` should be migrated to the insights, that's also great. I just want everyone on this thread to know that neither @jdolitsky nor I feel any territoriality about `rumble` and its functionality. Whatever is good for the engineering team and Chainguard is good to me. Period! :) You can kill `rumble` 🔪 . That's fine, but do let the downstream consumers know...

>  I really appreciate all of the work that you've done on this issue but we should take into account the remaining effort/time before continuing to do more.

Creating a unique CVEs graph could be done in less than two hours. But since this metric might not even be useful, I'll hold off.

> How many CVEs are in the images I just bought from Chainguard? (i.e., as measured by Grype but at some point it would be great to know the count from all of the scanners.

I think the existing "total non-unique vuln findings for Chainguard Images over last thirty days" graph covers this.

> How many of these CVEs are not being tracked by Chainguard?

Now you're in @luhring territory. Anything I would say here would be just plain dumb and ignorant compared to what @luhring has to say.

**Comment by jspeed-meyers** [comment]
Side note: It could be best to close this issue and start a new issue, either in `rumble` or `internal` :)

**Comment by rawlingsj** [comment]
So my thinking is to avoid disrupting any ongoing work I could have a spike today and see if I can get rumble deployed using the Insights pub/sub architecture.  

The main benefits would be:
- one GCP project where data related to images is ingested and queryable across many tables
- we can have monitoring and alerting setup which helps productionize our data based solutions
- leverage the separation of concerns for collecting data and ingesting into BigQuery OR allow other systems (i.e. Registry or Enforce) in the future to subscribe to these Topics and ingest data into their own data models for their own purposes.
- Insights codebases are private but we could `go import` public rumble codebase to reuse what's there already.

Let's check in tomorrow when US folks are back and see how far I get and if it's worth continuing or pausing consolidating Rumble and Insights work.  FWIW Insights is mainly about the the deployment and making it easy to extend, maintain, scale
and support.  My hope is the spike today will demonstrate these advantages while not impacting ongoing work or existing usecases that are relying on Rumble data today.
